agent/rpc/middleware/interceptors_test.go

```diff
@@ -0,0 +1,126 @@
+package middleware
+
+import (
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/require"
+)
+
+// obs holds all the things we want to assert on that we recorded correctly in our tests.
+type obs struct {
+	key    []string
+	start  time.Time
+	labels []metrics.Label
+}
+
+// recorderStore acts as an in-mem mock storage for all the RequestRecorder.Record() recorderFunc calls.
+type recorderStore struct {
+	lock  sync.Mutex
+	store map[string]obs
+}
+
+func (rs *recorderStore) put(key []string, o obs) {
+	rs.lock.Lock()
+	defer rs.lock.Unlock()
+
+	actualKey := strings.Join(append(key, o.labels[0].Value), "")
```

Comment:
the `key` here is always going to be the metric name, i.e. `rpc_server_call`. We need something else to use the map properly.

So we're making `rpc["" ]server[""]call[""]A.B" or whatever the method name is.

Curious if others have a better idea about this one.

Reply:
what about making it a stack instead of a map so you can also check the expected order if you have a use case with multiple calls?

Reply:
Hm, open to doing that. But expected order should not matter 🤔 

I don't think ordering is guaranteed by the underlying implementation of `go-metrics`. I think it's just a side effect.

Reply:
I agree in this use case it's not useful. As long as the key is something unique enough but can be guessed it should be fine.